tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,581 @@
+    self._gradients_applied = False
+    self._total_num_replicas = total_num_replicas
+    self._tokens_per_step = max(total_num_replicas, replicas_to_aggregate)
+    self._global_step = None
+    self._sync_token_queue = None
+
+    # The synchronization op will be executed in a queue runner which should
+    # only be executed by one of the replicas (usually the chief).
+    self._chief_queue_runner = None
+
+    # Remember which accumulator is on which device to set the initial step in
+    # the accumulator to be global step. This list contains list of the
+    # following format: (accumulator, device).
+    self._accumulator_list = []
+    # ma_run_hook should be called before SyncReplicasHook
+    self._ma_run_hook = False
+    # name: string. Name of the global variables and related operation on ps.
+    self._name = 'modelAverage'
+
+    self.add_variables_for_betweenGraph()
+
+  def add_variables_for_betweenGraph(self):
+    # change all variables to local variables
+    for v in variables.global_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of any name for'
+                             ' ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      ops.add_to_collection(ops.GraphKeys.LOCAL_VARIABLES, v)
+    # clear global_variables list
+    ops.get_default_graph().clear_collection(ops.GraphKeys.GLOBAL_VARIABLES)
+    # generate new global variables dependent on trainable variables
+    i = 0
+    for v in variables.trainable_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of a name for '
+                             'any ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      v_g = variable_scope.get_variable(name='%s_g%d' % (self._name, i),
+                                  initializer=v.initialized_value(),
+                                  trainable=False,
+                                  collections = [ops.GraphKeys.GLOBAL_VARIABLES,
+                                                 'global_model'])
+      v_block_grad = variable_scope.get_variable(
+                  name='%s_block_grad%d' % (self._name, i), shape=v.get_shape(),
+                  initializer=init_ops.constant_initializer(0.0, dtype=v.dtype),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'block_grads'])
+      v_nesterov = variable_scope.get_variable(
+                  name='%s_nesterov%d' % (self._name, i),
+                  initializer=v_g.initialized_value(),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'nesterov'])
+      i += 1
+    self._super_global_step = variables.Variable(0,
+                  name="%s_global_step" % self._name, trainable=False)
```

Comment:
Super_global_step counts steps per "model average”. So global_step is “interval_steps” times as many as super_global_step.